native/spark-expr/src/string_funcs/regexp_extract.rs

```diff
@@ -0,0 +1,452 @@
+            );
+        }
+
+        let subject = &args.args[0];
+        let pattern = &args.args[1];
+        let idx_val = if args.args.len() == 3 {
+            match &args.args[2] {
+                ColumnarValue::Scalar(ScalarValue::Int32(Some(i))) => *i as usize,
+                _ => {
+                    return exec_err!("regexp_extract_all idx must be an integer literal");
+                }
+            }
+        } else {
+            // Using 1 here to align with Spark's default behavior.
+            1
+        };
+
+        // Pattern must be a literal string
+        let pattern_str = match pattern {
+            ColumnarValue::Scalar(ScalarValue::Utf8(Some(s))) => s.clone(),
+            _ => {
+                return exec_err!("regexp_extract_all pattern must be a string literal");
+            }
+        };
+
+        // Compile regex once
+        let regex = Regex::new(&pattern_str).map_err(|e| {
+            internal_datafusion_err!("Invalid regex pattern '{}': {}", pattern_str, e)
+        })?;
+
+        match subject {
+            ColumnarValue::Array(array) => {
+                let result = regexp_extract_all_array(array, &regex, idx_val)?;
+                Ok(ColumnarValue::Array(result))
+            }
+            ColumnarValue::Scalar(ScalarValue::Utf8(s)) => {
+                match s {
+                    Some(text) => {
+                        let matches = extract_all_groups(text, &regex, idx_val)?;
+                        // Build a list array with a single element
+                        let mut list_builder =
+                            arrow::array::ListBuilder::new(arrow::array::StringBuilder::new());
+                        for m in matches {
+                            list_builder.values().append_value(m);
+                        }
+                        list_builder.append(true);
+                        let list_array = list_builder.finish();
+
+                        Ok(ColumnarValue::Scalar(ScalarValue::List(Arc::new(
+                            list_array,
+                        ))))
+                    }
+                    None => {
+                        // Return NULL list using try_into (same as planner.rs:424)
+                        let null_list: ScalarValue = DataType::List(Arc::new(
+                            arrow::datatypes::Field::new("item", DataType::Utf8, false),
+                        ))
+                        .try_into()?;
+                        Ok(ColumnarValue::Scalar(null_list))
+                    }
+                }
+            }
+            _ => exec_err!("regexp_extract_all expects string input"),
+        }
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+}
+
+// Helper functions
+
+fn extract_group(text: &str, regex: &Regex, idx: usize) -> Result<String> {
+    match regex.captures(text) {
+        Some(caps) => {
+            // Spark behavior: throw error if group index is out of bounds
```

Comment:
AFAIS Spark returns an empty string when there is no such group - https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L809-L821